jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java

```diff
@@ -0,0 +1,193 @@
+import java.time.Duration;
+import java.time.Instant;
+import java.time.format.DateTimeParseException;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+/** Checks if Jib is up-to-date. */
+public class UpdateChecker {
+
+  /** JSON template for the configuration file used to enable/disable update checks. */
+  @VisibleForTesting
+  static class ConfigJsonTemplate implements JsonTemplate {
+    private boolean disableUpdateCheck;
+
+    @VisibleForTesting
+    void setDisableUpdateCheck(boolean disableUpdateCheck) {
+      this.disableUpdateCheck = disableUpdateCheck;
+    }
+  }
+
+  /**
+   * Begins checking for an update in a separate thread.
+   *
+   * @param versionUrl the location to check for the latest version
+   * @return a new {@link UpdateChecker}
+   */
+  public static UpdateChecker checkForUpdate(String versionUrl) {
+    ExecutorService executorService = Executors.newSingleThreadExecutor();
+    Future<Optional<String>> messageFuture =
+        executorService.submit(
+            () ->
+                performUpdateCheck(
+                    Verify.verifyNotNull(ProjectInfo.VERSION), versionUrl, getConfigDir()));
+    executorService.shutdown();
+    return new UpdateChecker(messageFuture);
+  }
+
+  @VisibleForTesting
+  static Optional<String> performUpdateCheck(
+      String currentVersion, String versionUrl, Path configDir) {
+    // Abort if offline or update checks are disabled
+    if (Boolean.getBoolean(PropertyNames.DISABLE_UPDATE_CHECKS)) {
+      return Optional.empty();
+    }
+
+    Path configFile = configDir.resolve("config.json");
+    Path lastUpdateCheck = configDir.resolve("lastUpdateCheck");
+    try {
+      if (Files.exists(configFile)) {
+        // Abort if update checks are disabled
+        ConfigJsonTemplate config =
+            JsonTemplateMapper.readJsonFromFile(configFile, ConfigJsonTemplate.class);
+        if (config.disableUpdateCheck) {
+          return Optional.empty();
+        }
+      } else {
+        // Generate config file if it doesn't exist
+        ConfigJsonTemplate config = new ConfigJsonTemplate();
+        Files.createDirectories(configDir);
+        try (OutputStream outputStream = Files.newOutputStream(configFile)) {
+          JsonTemplateMapper.writeTo(config, outputStream);
+        }
+      }
+
+      // Check time of last update check
+      if (Files.exists(lastUpdateCheck)) {
+        String fileContents =
+            new String(Files.readAllBytes(lastUpdateCheck), StandardCharsets.UTF_8);
+        try {
+          Instant modifiedTime = Instant.parse(fileContents);
+          if (modifiedTime.plus(Duration.ofDays(1)).isAfter(Instant.now())) {
+            return Optional.empty();
+          }
+        } catch (DateTimeParseException ex) {
+          // Ignore parse failure; assume update check hasn't been performed
```

Comment:
Following @loosebazooka's suggestion, getting here most likely means the file is corrupted.

Reply:
I think we can auto-delete this one. The user probably shouldn't be touching `lastUpdateCheck`.